[2단계 - 로또 리팩토링] 멍구(이유영) 미션 제출합니다.

yuyoungrhee opened the pull request:
이프 안녕하세요! 1단계에서 많은 코멘트 주신 덕분에 나름 많은 고민을 한 것 같아요. 관련해서 해결한 고민과, 아직 남아있는 고민을 정리했습니다.

### 클래스 설계
![IMG_BC0A8BDED104-1](https://github.com/user-attachments/assets/47c44e1c-8e56-4074-9269-8f73f4639c78)

## 주요 변경사항 및 코멘트 반영 결과

### Dto 사용에 대하여
> 데이터 전송을 위해 왜 Dto를 활용할까요? 도메인 객체를 주고 받으면 안되는 걸까요??

관련해서 고민하다보니 도메인 객체를 주고받는 게 더 자연스럽다고 생각했습니다.

Dto는 보통 계층 간 데이터 전송을 위해서 사용하게 되는데, 각 계층별로 필요한 직접 도메인을 주고 받으면 도메인에 각 계층이 영향을 받기 때문입니다. 계층 간 결합을 낮추기 위한 객체라고 생각해요!

그런데 다시 고민해보면서, 레벨1의 자바 미션에서는 보통 도메인 내부 협력을 통해 기능을 처리하고 있습니다. 따라서 도메인 내부 협력을 목적으로 Dto가 쓰이고 있고, 그렇다면 그냥 도메인 객체를 주고받는 것이 더 자연스러운 설계 같다고 결론 내렸습니다.

변경 결과: WinningInform을 Dto에서 일반 도메인 객체로 변경하였고 관련 검증 로직도 InputController가 아니라 도메인 내부에서 진행하도록 변경했습니다.

### Lotto 도메인 내부 정렬 테스트 관련
~~Lotto 필드인 numbers에 대한 getter가 열려있지 않기 때문에 toString()을 통한 테스트를 진행했습니다.~~

```java
@DisplayName("로또 번호가 오름차순으로 정렬되어야 한다.")
    @Test
    void testSort_lottoNumber() {
        // given
        List<Integer> list = new ArrayList<>(List.of(3, 5, 4, 1, 6, 2));
        Lotto lotto = new Lotto(list);

        List<Integer> sortedList = list.stream().sorted().toList();

        //assertThat(lotto.toString()).isEqualTo(sortedList.toString());
        assertThat(lotto.getNumbers()).isEqualTo(sortedList);
    }
```

~~그런데 toString을 통한 테스트가 좋은 테스트일지 확신이 없습니다. 만약 객체 출력을 위한 포맷 변경을 하게 된다면, 내부 정렬은 되어있어도 테스트가 실패할 수 있을 것 같습니다. 테스트를 위한 protected getter를 열어두는 게 더 좋을까요?~~
출력을 위한 getter를 열어두게 되면서 테스트도 getter를 통해 진행하게 됐습니다. 지금은 테스트만을 위한 getter는 아니지만, 이프는 테스트를 위한 getter에 대해 어떻게 생각하는 지 궁금합니다. 정렬 테스트 같은 경우 공개 인터페이스를 통하여 테스트를 할 수 있는 수단이 없어 public 메서드만으로는 정렬이 되었는지 테스트할 수 없었습니다.

### 도메인에 남아있는 View 관련 로직들
이프의 공통 피드백을 통해 View와 Controller에 대해서 고민해보고, 또 입출력이 없었다면 어떻게 코드를 설계했을까에 대하여 생각했습니다.

그러다 보니 자연스럽게 도메인에 남아있는 출력 관련 toString 로직에 대해서 고민하게 되었는데요, 결론적으로 도메인에서 출력을 위한 toString 설계는 좋은 방법이 아니다라고 판단했습니다. toString은 본래 디버깅이나 로깅을 위한 역할이기 때문에 View에 맞는 포맷을 위해 toString을 사용하지 않아야겠다고 생각했습니다.

1단계 리팩토링 전 Dto들은 실제로 도메인과 도메인 사이의 데이터전송을 위한 Dto였기 때문에 Dto를 삭제하고 도메인으로 변경하는 방식으로 리팩토링을 진해왰었는데, 그렇다면 **Dto를 여기서 사용하면 되는 걸까요?** 이렇게 사용한다면, 도메인 -> View 를 위한 데이터 전송 객체의 의미를 가질 수 있을까요?

### enum 클래스인 MatchResult 설계 관련

> MatchResult를 인자로 넘겨받으면 검증에 대한 책임을 MatchRank가 지니지 않아도 되지 않을까요??
> enum 클래스가 커진다면 책임 분리에 문제가 있는 것이기에 도메인 설계를 다시 짚어보는 것도 좋습니다!

그 생각을 못했습니다.!!! MatchResult를 인자로 넘겨받으면 굳이 LottoStatistics에서 result에 대한 값들을 꺼내서 다시 인자로 넘겨줄 필요도 없어 더 간단해집니다. 왜 이 생각을 못했는지😅 관련해서 수정했습니다! 일치 개수에 대한 검증도 MatchResult가 하는 것으로 옮겼습니다.

### Money의 isLessThanLottoPrize 메서드명
> `isLessThanLottoPrize`라는 메서드명으로 수정하셨군요!
> 이 경우는 LOTTO_PRIZE 상수명 변경에 영향을 받진 않을까 조심스레 생각해보아요~

`isNotEnough()`라는 메서드명으로 변경했습니다. 로또 하나를 구매할 수 없을 만큼 돈이 부족한 상태를 나타내는 메서드명을 생각하면서 고민해보았는데 마땅히 적당한 메서드명이 떠오르지 않았어요🥲


## 남은 고민 + 남은 코멘트
### 랜덤 테스트
> 멍구에게 랜덤 확률을 조정하는 기획이 추가된다면, 현재 코드에서 어떻게 적용할 수 있을까요?
> OCP를 지키면서 유연한 설계를 해보세요! 그럼 자연스레 테스트도 가능한 구조가 탄생해요.

랜덤 확률을 조정하는 기획이 추가된다면, LottoNumberGenerator라는 인터페이스, 그리고 “랜덤하게” 숫자를 뽑는 RandomNumberGenerator라는 구현체, 그리고 커스텀확률로 숫자를 뽑는 CustomNumberGenerator로 나누어서 구현을 할 것 같습니다. 그럼 자연스럽게 테스트 코드에서도 테스트에 필요한 NumberGenerator를 자유롭게 구성할 수 있어서 랜덤 값을 테스트할 수 있을 것 같아요.

다만 이프의 원래 질문인 랜덤 값이 1~45 범위에서 생성된다는 것을 어떻게 테스트로 커버할 수 있을까? 에 대해서 아직 해결하지 못했습니다. 결국 테스트 해야하는 건 여전히 RandomNumberGenerator라는 랜덤 관련 구현체여서 같은 고민이 반복되는 것 같아요... 제가 잘못 접근하고 있는 걸까요?

### 상수 관리
> 도메인이 지닌 상수를 꼭 private으로 정의해야 하나요?

도메인 캡슐화 때문에 private으로 정의해야한다고 막연히 생각했습니다. 도메인 내부 규칙이 외부에 노출되게 되는데 상수는 불변이기 때문에 노출되어도 상관 없을까요?
또, 로또 번호 개수(6개)는 로또 번호를 개수대로 생성하는 번호생성기에서 사용하고 있기도 한데요, 이처럼 여러곳에서 사용되는 상수라면 도메인 규칙에 관련되어있지만 공통으로 관리하는 것이 좋을까요? 현재 상수 클래스에서 관리하고 있는 모든 상수가 Lotto 말고도 다른 클래스에서 사용하고 있습니다(검증 등)

### DependencyConfig 의존성 관리
> 의존성 주입이 열리면 어떤 장점이 있을까요?

각 객체가 서로 구체적인 구현에 의존하지 않으면서 결합도가 낮아지는 것이 가장 큰 장점인 것 같습니다.
그렇지만 이건 제가 DependencyConfig를 사용하여 얻은 장점은 아닌 거 같아요.. DependencyConfig는 의존성 주입을 돕는 게 아니라 관리를 도와주는 역할인 것 같아요.

yuyoungrhee committed:
refactor(ErrorMessage): 유효하지 않은 숫자 입력 예외메시지 변경

yuyoungrhee committed:
refactor(WinningInform): Dto가 아닌 도메인 객체로 변경

WinningInform이 Dto가 아닌 도메인 객체로써 보너스번호와 당첨번호에 대해서 검증하고 협력에 참여하도록 수정함

yuyoungrhee committed:
test(Lotto): 정렬 테스트 추가

toString을 통한 정렬 여부 확인 테스트 추가

yuyoungrhee committed:
refactor(MatchRank): 메서드 인자 설정 변경 및 일치 개수에 대한 검증 책임 변경

- 일치 개수와 보너스 여부가 아닌 MatchResult 자체를 인자로 넘겨받도록 변경함
- 일치 개수에 대한 검증은 MatchResult가 가지도록 책임을 옮김

yuyoungrhee committed:
refactor(Money): 메서드명 변경

- LOTTO_PRIZE라는 상수명에 의존하는 메서드명을 변경

yuyoungrhee committed:
refactor(Money): record 활용

Money 객체는 값 그 자체로서의 의미를 갖고, 상태 변경 없이 단순 데이터와 간단한 검증 로직만 포함하기 때문에 Record를 활용함

yuyoungrhee committed:
refactor: 도메인에 존재하는 출력 관련 로직들 삭제 -> Dto 사용

도메인의 toString() 오버라이드 메서드를 View의 출력을 위해 사용하지 않도록 변경
- Dto를 사용하여 전달할 수 있도록 수정
- Lotto 도메인의 toString()은 디버깅을 위해 남겨둠

yuyoungrhee committed:
refactor(test): 테스트 일부 패키지 이동

잘못된 패키지 이동

sinb57 commented on src/main/java/lotto/domain/WinningInform.java:
```diff
@@ -0,0 +1,36 @@
+        validateBonusDuplicate(winningLotto, bonusNumber);
+    }
+
+    private static void validateBonusRange(int bonus) {
+        if (bonus < LOTTO_RANGE_MINIMUM || bonus > LOTTO_RANGE_MAXIMUM) {
+            throw new IllegalArgumentException(ERROR_BONUS_OUT_OF_RANGE);
+        }
+    }
```
Lotto에도 동일한 검증로직이 있네요! 정책에 변경이 생긴다면, 여러 곳을 수정해야 해요~ 어떻게 하면 중복을 제거할 수 있을까요??

yuyoungrhee replied:
로또 숫자도 객체로 포장하면서 해결할 수 있었습니다! 범위에 대한 검증은 이제 로또 숫자 객체인 LottoNumber 안에서 수행하게 됩니다.

sinb57 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -91,4 +91,9 @@ private static void sortNumbers(List<Integer> numbers) {
     }
+
+    public List<Integer> getNumbers() {
+        return Collections.unmodifiableList(numbers);
```
Collections.unmodifiableList()으로 감싼 이유가 궁금해요~ㅎ

yuyoungrhee replied:
Lotto의 내부 필드를 외부에서 수정하지 못하도록 하기 위함이었습니다.
getNumbers는 테스트에서 사용되고 있었는데요,(현재는 정렬 테스트 삭제로 사용되고있지 않습니다) 사진과 같이 test에서 리스트를 수정할 수 있습니다. 따라서 읽기 전용으로 numbers를 반환하도록 했어요. 테스트, 그리고 Dto 등에서도 수정이 아니라 읽는 기능만 필요해서요!

<img width="554" alt="스크린샷 2025-02-16 오후 11 12 59" src="https://github.com/user-attachments/assets/dca24a7e-04e6-4433-be7c-8672af9ab65d" />

sinb57 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -91,4 +91,9 @@ private static void sortNumbers(List<Integer> numbers) {
     public String toString() {
         return numbers.toString();
     }
+
```
Lotto 에 대한 피드백 남겨보아요~

- 수직거리를 고려해주세요
validate 메서드를 보면 Lotto 생성자에서 사용되지만, 생성자와 validate 메서드의 사이에 다른 메서드가 배치되어 있어요. 이 경우 생성자에서 시작하여 호출 메서드를 들여다보기 위해서는 비교적 먼 거리를 이동하여 메서드를 살펴보게 됩니다. 크진 않지만 시간과 노력이 소모되다 보니, 수직거리를 고려해주시면 좋아요!
시간 괜찮을 때 "클린코드(p.101)"를 읽어보세요~

- 정렬 필요성
생성자를 보면 sortNumbers가 존재하네요. 해당 객체가 "정렬"된 로또 숫자를 가지고 있어야 하나요? 요구사항에서 "정렬"은 무엇을 위한 행위일까요?

- containsNumber 메서드 시그니처
`public boolean containsNumber(int number)` 메서드의 시그니처를 보면 메서드명과 인자명에 중복된 정보가 존재해요. `public boolean contains(int number)` 라고 해도 충분히 의미 전달이 되는데 어떻게 생각하시나요?

yuyoungrhee replied:
**1. 수직거리**
private 메서드는 public 메서드보다 아래에 위치해야한다고 생각했는데 연관성에 따라 형식을 맞춰야 했군요!
> 한 함수가 다른 함수를 호출한다면 두 함수는 세로로 가까이 배치한다.

validate 메서드를 먼저 배치하도록 변경했습니다!

**2. 정렬 필요성**
로또가 자신의 번호를 정렬해야한다고 생각했는데, 요구사항에 따르면 정렬은 "출력"을 위한 행위군요. 코드를 작성하면서도 출력을 해보고 나서 정렬 기능이 필요하다고 생각했습니다. 그러니 도메인을 변경하는 것이 아니라 "출력"을 담당하는 view에서 처리하도록 해야됐네요..!

+ 추가로 로또의 정렬 기능을 삭제하면서 정렬 테스트도 필요 없어졌는데, 왜 정렬 테스트가 어려웠는지 알게 되었어요. public 메서드를 통해 내부 구현을 검증하는 것이 좋은데, 정렬 기능은 로또의 요구사항이 아니니 외부에서 정렬 기능을 요구하거나 사용하지 않았습니다. 따라서 테스트를 하기 위해서는 getter라든가 toString()을 통한 내부 검증이 필요했던 거예요! 이제는 그럴 필요가 없어졌습니다.

3. containsNumber의 메서드 시그니처
동의합니다! 이미 맥락이 명확하기 때문에 contains만 해도  좋을 것 같아요. 앞으로도 중복되지 않도록 필요한 정보를 담는 메서드명을 고민해볼게요..!

sinb57 commented on src/main/java/lotto/domain/Money.java:
```diff
@@ -3,18 +3,20 @@
+    public Money {
+        validate(money);
+    }
+
+
```
intellij의 코드 포매팅 기능 사용을 습관화하시면 좋아요!

yuyoungrhee replied:
수정했습니다!

sinb57 commented on src/main/java/lotto/domain/Money.java:
```diff
@@ -36,7 +38,4 @@ public int getAmount() {
         return money / LOTTO_PRIZE;
```
```java
    public List<Lotto> buyLottos(Money money) {
        List<Lotto> lottos = new ArrayList<>();
        for (int i = 0; i < money.getAmount(); i++) {
            List<Integer> numbers = lottoNumberGenerator.generate();
            lottos.add(new Lotto(numbers));
        }
        return lottos;
    }
```
`getAmount` 관련하여 질문 드려요~ `money.getAmount()`를 보았을 때 느낀 점은 "금액을 가져온다"였어요. 그런데 내부는 로또 구매 개수를 반환하도록 구현하셨더라고요. 이를 간단히 수정한다면 메서드명을 변경하면 되겠지만, 그에 앞서 Money 객체의 역할을 생각해보면 좋겠어요~ 제가 생각하는 역할과는 다르겠지만, Money라는 이름만 봤을 때는 "돈" 그자체에 대한 원시값으로 보여요. 그러다보니 Money가 로또 구매 개수를 계산하는 책임을 지는 것이 옳은지 잘 모르겠네요

yuyoungrhee replied:
Amount가 "개수"를 의미하는 줄 알았는데 찾아보니 "금액"이라는 뉘앙스에 더 가깝군요.. 영어 실력 부족 이슈입니다😔

더해서 Money의 역할과, 그리고 상수의 위치에 대해서 고민하다보니 이 코멘트도 같은 고민을 공유하더라고요! LOTTO_PRIZE라는 상수는 어떤 도메인과 연관이 있을까? 고민하다보니 돈을 받고 로또를 개수만큼 발급해주는 LottoShop 도메인에서 관리해야겠다는 생각이 들었습니다. 따라서 **로또 구매 개수를 계산하는 책임은 LottoShop이 가지도록 변경**했어요.

sinb57 commented on src/main/java/lotto/domain/Wallet.java:
```diff
@@ -10,26 +11,17 @@ public Wallet(List<Lotto> lottos) {
     }
 
-    public List<MatchResult> getMatchResults(Lotto matchLotto, int bonus) {
+    public List<MatchResult> getMatchResults(WinningInform winningInform) {
```
Wallet은 로또 묶음을 지니고 있고 당첨 결과 계산만을 책임지고 있네요?
이름을 보면 Money의 역할을 여럿 들고와도 괜찮아보이고, lottoShop의 책임을 가져와도 좋아보이네요

yuyoungrhee replied:
어떤 식으로 책임을 분리해야할지 감이 잘 안 오네요. LottoShop의 책임을 가져오기에는 자신의 필드인 List<Lottos>를 스스로 생성하게 되는데 이렇게 해도 괜찮은 설계일까요?

sinb57 replied:
꼭 lottos를 가지고 있어야 할까요?? wallet은 lottos 보단 money를 가지는게 이름에 걸맞지 않나 싶어요.

sinb57 commented on src/main/java/lotto/domain/MatchRank.java:
```diff
@@ -34,29 +32,24 @@ public String getMatchData() {
         }
 
         for (MatchRank statistics : values()) {
-            if (statistics.number == matchNumber) {
+            if (statistics.number == matchCount) {
                 return statistics;
             }
         }
```
MatchRank 타입으로 statistics 이름이 정확한 의미 전달을 하고 있는지 궁금합니다!
`return statistics;`보고 "통계를 반환?" 헷갈렸어요~ㅎ

yuyoungrhee replied:
수정했습니다! 원래 클래스명이 MatchStatistics라서 미처 수정하지 못했습니다.

sinb57 commented on src/main/java/lotto/domain/MatchRank.java:
```diff
@@ -34,29 +32,24 @@ public String getMatchData() {
 
-    public static MatchRank getMatchRank(int matchNumber, boolean bonus) {
-        validateMatchNumber(matchNumber);
+    public static MatchRank getMatchRank(MatchResult matchResult) {
```
> > enum 클래스인 MatchResult 설계 관련
MatchResult를 인자로 넘겨받으면 검증에 대한 책임을 MatchRank가 지니지 않아도 되지 않을까요??
enum 클래스가 커진다면 책임 분리에 문제가 있는 것이기에 도메인 설계를 다시 짚어보는 것도 좋습니다!
>
> 그 생각을 못했습니다.!!! MatchResult를 인자로 넘겨받으면 굳이 LottoStatistics에서 result에 대한 값들을 꺼내서 다시 인자로 넘겨줄 필요도 없어 더 간단해집니다. 왜 이 생각을 못했는지😅 관련해서 수정했습니다! 일치 개수에 대한 검증도 MatchResult가 하는 것으로 옮겼습니다.

당시 "matchCount가 -1이 나온다거나 로또번호 개수인 6개를 넘어가는 경우"를 어떻게 처리하는 것이 좋을지 질문 주셨고, 이어지는 내용에서는 "MatchStatistics(현재 MatchRank)가 매칭까지 처리하게 되어 enum 클래스가 너무 커질 수 있겠다는 우려"가 있다고 하셔서 "MatchResult를 인자로 넘겨받아 처리"하는 것을 제안드려보았어요ㅎ

하지만 위 방법이 최선은 아니라고 생각하는데요. 우선 고민하셨던 예외 케이스는 로직상 발생 가능성이 없어요. 예외를 발생하는 것은 좋은 방법이지만, 무조건 예외를 발생해야 한다고 생각하지는 않는데요. 비정상 케이스는 모두 NO_MATCH로 기본 처리해도 괜찮지 않나 싶어요.

이렇게 하면 로직이 간단해질 것으로 보이는데요. 우선 검증 로직을 담고 있는 MatchResult의 존재 의의를 다시 살펴볼 수 있고, MatchResult와 연관된 객체들을 건드리며 관계를 단순화할 수 있어요.

추상적인 제안이라 적용하기 어렵겠지만, DTO의 역할을 맡고 있는 MatchResult가 정말 필요한지 생각해보셨으면 해서 남깁니다.

yuyoungrhee replied:
도메인의 경우 자기 자신에 대해서 어디까지 책임져야하는지에 대해서, 협력 관계에서의 객체 말고도 자기 자신을 온전히 책임져야하는 지 궁금했습니다. 질문드렸던 발생 가능성이 없는 예외에 대해서도 같은 맥락으로 의문이 들었습니다. 그러나 지금은 예외가 생겨도 NO_MATCH가 반환되는 형식이니 예외의 경우에도 도메인에 문제가 생기지는 않을 것 같습니다..

#### MatchResult의 책임 필요성
현재 MatchResult가 필요하지 않다는 것에 동의합니다. Dto가 아닌 Domain의 역할을 수행하고 있는 줄 알았는데, matchCount, isBonusMatched라는 Lotto에서 넘어온 값을 저장하고 있는 형태라 값 저장소의 역할을 하고 있는 것 같습니다.. 
수정: Lotto 에서 일치 개수와 보너스번호 일치 여부를 판단하여 **MatchRank를 반환하도록 했습니다.** Lotto 스스로 자신이 어떤 Rank인지 확인하여 반환하는 역할을 하게 되었습니다.

sinb57 replied:
> 도메인의 경우 자기 자신에 대해서 어디까지 책임져야하는지에 대해서, 협력 관계에서의 객체 말고도 자기 자신을 온전히 책임져야하는 지 궁금했습니다. 질문드렸던 발생 가능성이 없는 예외에 대해서도 같은 맥락으로 의문이 들었습니다. 그러나 지금은 예외가 생겨도 NO_MATCH가 반환되는 형식이니 예외의 경우에도 도메인에 문제가 생기지는 않을 것 같습니다..

도움이 될지는 모르겠으나, "오브젝트 책의 부록A. 계약에 의한 설계"를 읽어보시면 좋을 듯 해요.

sinb57 commented on src/test/java/lotto/domain/WinningInformTest.java:
```diff
@@ -0,0 +1,35 @@
+        }).isInstanceOf(IllegalArgumentException.class)
+                .hasMessage(ERROR_DUPLICATE_WINNING_AND_BONUS);
+    }
+
```
WalletTest의 testToString 테스트와 같이 불필요한 테스트는 제거 부탁드리고,
가능하면 객체들의 열린 메서드들은 모두 테스트 작성해주세요~

sinb57 commented on src/test/java/lotto/domain/LottoTest.java:
```diff
@@ -50,4 +50,16 @@ void test_validateDuplicate() {
+
+        assertThat(lotto.getNumbers()).isEqualTo(sortedList);
+    }
+
```
LottoShopTest에 피드백 하나 드리자면, 아래의 경우 hasSize라는 친구를 사용하시면 좋아요.
```
        assertThat(lottos.size()).isEqualTo(money.getAmount());
```

yuyoungrhee replied:
감사합니다! 수정했어요. 알고있는 메서드였는데 활용하지 못했습니다..

sinb57 commented on src/test/java/lotto/domain/LottoStatisticsTest.java:
```diff
@@ -2,59 +2,84 @@
-        Wallet wallet = new Wallet(List.of(
-                new Lotto(List.of(1, 2, 3, 7, 8, 9))
-        ));
+    private static Stream<LottoStatisticsTestData> provideTestData() {
+        // 케이스 1: 모든 MatchRank가 한개씩 있는 경우
+        Lotto winningLotto = new Lotto(List.of(1, 2, 3, 4, 5, 6));
+        int bonus = 7;
+        WinningInform winningInform = new WinningInform(winningLotto, bonus);
+
+        Lotto ticket1 = new Lotto(List.of(1, 2, 3, 7, 8, 9)); //3개 매칭
+        Lotto ticket2 = new Lotto(List.of(1, 2, 3, 4, 7, 8)); // 4개 매칭
+        Lotto ticket3 = new Lotto(List.of(1, 2, 3, 4, 5, 8)); // 5개 매칭
+        Lotto ticket4 = new Lotto(List.of(1, 2, 3, 4, 5, 7)); // 5개 매칭 + 보너스
+        Lotto ticket5 = new Lotto(List.of(1, 2, 3, 4, 5, 6)); // 6개 매칭
+        Wallet wallet = new Wallet(List.of(ticket1, ticket2, ticket3, ticket4, ticket5));
+
+        Map<MatchRank, Integer> expectedCounts1 = new EnumMap<>(MatchRank.class);
+        expectedCounts1.put(MatchRank.MATCH_THREE, 1);
+        expectedCounts1.put(MatchRank.MATCH_FOUR, 1);
+        expectedCounts1.put(MatchRank.MATCH_FIVE, 1);
+        expectedCounts1.put(MatchRank.MATCH_BONUS, 1);
+        expectedCounts1.put(MatchRank.MATCH_SIX, 1);
+        expectedCounts1.put(MatchRank.NO_MATCH, 0);
```
테스트 인자를 만들기 위해 로직이 생겼는데, 이 로직은 테스트가 불필요한가요? 테스트 인자 생성이 잘못되어 실패해야 하는 테스트가 성공하면 어떡하죠?

sinb57 commented on src/main/java/lotto/domain/MatchRank.java:
```diff
@@ -10,8 +10,6 @@ public enum MatchRank {
 
-    private static final int MIN_MATCH_NUMBER = 0;
-    private static final int MAX_MATCH_NUMBER = 6;
     private final static String BONUS_OUTPUT_MESSAGE = "%d개 일치, 보너스 볼 일치 (%d원)- ";
     private final static String BASIC_OUTPUT_MESSAGE = "%d개 일치 (%d원)- ";
```
출력형식이 수정되면 MatchRank 객체를 수정해야 하는데요. 멍구는 domain과 view를 어떤 목적으로 어떻게 나누었나요??

yuyoungrhee replied:
이번 리팩토링 과정에서 출력 로직을 도메인에서 가지고 있다면 출력 형식이 수정됐을 때 도메인을 변경하는 방식이 이상하다고 답변 드렸는데, MatchRank 객체에 있는 출력 로직을 확인 못하고 남겨두었습니다.. 수정했습니다!

추가로 저는 Domain에서는 비즈니스 로직만 출력하고, View에서는 도메인 객체에 있는 값을 요구사항의 형태로 가공해서 출력하는 역할로 나누었습니다. 1단계에서는 Domain에서 출력의 형태까지 정하도록 하였는데, 편의성을 위해서 코드를 짰지만 단일책임 원칙이 지켜지지 않았던 것 같아요. 2단계에서는 MatchRank까지 수정하면서 도메인에 있는 출력 관련 로직을 없애고 view에게 책임을 옮겼습니다.

sinb57 commented on src/main/java/lotto/controller/InputController.java:
```diff
@@ -1,14 +1,10 @@
+import lotto.domain.WinningInform;
 import lotto.view.InputView;
 
 public class InputController {
```
InputController는 InputView와 어떠한 목적으로 구분지었나요?
그리고 InputView는 현재 Scanner의 역할만 수행하고 있고 다른 기능이 없는데, InputView는 존재 의의가 무엇인가요?

yuyoungrhee replied:
inputView는 사용자의 입력을 받아오는 역할이라고 생각합니다. 단순 UI 계층으로 생각해서 원본을 가공하지 않고 그대로 Controller로 전달하는 방법을 선택했습니다. 이렇게 할 경우 장점은 view가 도메인 로직과 분리된다는 점입니다.

그래서 input을 도메인에서 원하는 형태로 바꾸기 위한 로직을 inputController에 담아 본 Controller에서 사용할 수 있도록 하였습니다.

이프는 inputView에서 비즈니스로직에서 요구하는 객체로의 변환 책임까지 맡아야한다고 생각하시나요?

sinb57 replied:
inputView에 대해 사용자의 입력을 받아오는 역할이라고 말씀주셨지만, 제가 보기에는 inputView가 Scanner와 차이가 없어보이기 때문에 오히려inputController가 사용자의 입력을 받아오는 역할을 수행한다고 보여요. 더불어 객체로의 변환까지도요.

view가 도메인 로직과 분리되도록 하기 위함에 대해 공감해요.
다만 앞서 질문을 드린 이유는 InputView의 역할이 불명확했기 때문이에요. read 메서드를 지금의 InputController에 private으로 선언하여 사용해도 문제 없어 보이거든요.

yuyoungrhee replied:
View의 역할이 불명확했다는 점에 동의합니다.. View의 역할을 어디까지 볼 것이냐에 대해서 명확하게 정리되지 않은 상태에서 코드를 작성했던 것 같아요. 

Controller는 View를 통해 "어떤 입력을 받을지" 결정하도록 하고, View의 세부적인 내용은 알지 못하도록 변경했습니다. View에서는 어떤 문구로 보여줄지 결정하고, 입력받을 자료형까지 결정합니다. 자료형을 통해 도메인 객체로 변환하는 역할은 InputController에서 하도록 했습니다 재시도 로직을 포함해서요!

sinb57 commented on src/main/java/lotto/domain/Money.java:
```diff
@@ -24,7 +26,7 @@ private void validateDivided(int money) {
         }
     }
 
-    private boolean isLessThanLottoPrize(int money) {
+    private boolean isNotEnough(int money) {
```
의미가 잘 전달되네요! 고민하느라 고생 많으셨어요~ 이제 다른 질문을 드릴게요ㅎ
`money < LOTTO_PRIZE` 비교는 왜 `isNotEnough` 메서드로 분리하셨나요??

yuyoungrhee replied:
if문 안의 조건을 좀 더 명확히 하기 위하여 메서드를 통해 분리하였는데, 사실 이정도까지의 분리는 필요 없었던 것 같습니다. 조건이 이해하기 어려운 형태였다면 분리하는 게 맞지만, 지금은 `validateEnoughMoney` 안에 `ifNotEnough`라는 조건 판단 메서드가 있는 것인데 validateEnoughMoney` 메서드명으로도 충분히 조건이 이해가 됐을 것 같아요. 

관련해서 다른 validate도 조건을 통합해서 하나의 메서드로 작동하도록 했습니다.

sinb57 commented on src/main/java/lotto/domain/Money.java:
```diff
@@ -3,18 +3,20 @@
 
-public class Money {
-    private final int money;
+public record Money(int money) {
```
원시값을 포장했듯 로또 숫자도 포장하면 어떨까요?

yuyoungrhee replied:
이번 리팩터링을 하면서 로또 숫자를 객체로 만들어야겠다는 필요성을 느꼈던 것 같습니다, 코멘트 주신 것처럼 LottoNumber를 객체로 만들었습니다.

sinb57 reviewed (changes requested):
안녕하세요 멍구~ 리팩터링 하느라 고생 많으셨어요! 코멘트 확인 부탁드려요

---

> 출력을 위한 getter를 열어두게 되면서 테스트도 getter를 통해 진행하게 됐습니다. 지금은 테스트만을 위한 getter는 아니지만, 이프는 테스트를 위한 getter에 대해 어떻게 생각하는 지 궁금합니다. 정렬 테스트 같은 경우 공개 인터페이스를 통하여 테스트를 할 수 있는 수단이 없어 public 메서드만으로는 정렬이 되었는지 테스트할 수 없었습니다.

테스트를 위해 getter를 선언하는 것에 대해 거부감을 가지고 있지 않지만, 되도록 하지 않으려고 해요ㅎ 멍구는 getter를 왜 열어두면 안된다고 생각하시나요? 말씀하셨듯 테스트를 작성할 때 편하지 않으셨나요??

> 1단계 리팩토링 전 Dto들은 실제로 도메인과 도메인 사이의 데이터전송을 위한 Dto였기 때문에 Dto를 삭제하고 도메인으로 변경하는 방식으로 리팩토링을 진해왰었는데, 그렇다면 Dto를 여기서 사용하면 되는 걸까요? 이렇게 사용한다면, 도메인 -> View 를 위한 데이터 전송 객체의 의미를 가질 수 있을까요?

Dto라는 이름대로 데이터 전송을 위해서라면 사용해도 괜찮아요. 다만 Dto를 언제 왜 사용하는지가 중요한데요. 앞서 드렸던 질문에 대해 "Dto는 보통 계층 간 데이터 전송을 위해서 사용하게 되는데, 각 계층별로 필요한 직접 도메인을 주고 받으면 도메인에 각 계층이 영향을 받기 때문입니다. 계층 간 결합을 낮추기 위한 객체라고 생각해요!"라고 말씀 주셨네요. 도메인에 각 계층이 영향을 받는다는 부분이 잘 이해되지 않는데 이부분 자세히 설명 주실 수 있나요?
계층 간 이동에 DTO를 사용하는 목적이라면 지금의 LottoDto와 WalletDto는 목적을 수행하고 있지만, 단순히 계층이 쪼개어 있기 때문에 DTO를 사용하는 것은 반대하고 싶어요. 지금 미션 규모에서는 DTO가 사실 필요없다고 생각하거든요.

> 랜덤 확률을 조정하는 기획이 추가된다면, LottoNumberGenerator라는 인터페이스, 그리고 “랜덤하게” 숫자를 뽑는 RandomNumberGenerator라는 구현체, 그리고 커스텀확률로 숫자를 뽑는 CustomNumberGenerator로 나누어서 구현을 할 것 같습니다. 그럼 자연스럽게 테스트 코드에서도 테스트에 필요한 NumberGenerator를 자유롭게 구성할 수 있어서 랜덤 값을 테스트할 수 있을 것 같아요.
>
> 다만 이프의 원래 질문인 랜덤 값이 1~45 범위에서 생성된다는 것을 어떻게 테스트로 커버할 수 있을까? 에 대해서 아직 해결하지 못했습니다. 결국 테스트 해야하는 건 여전히 RandomNumberGenerator라는 랜덤 관련 구현체여서 같은 고민이 반복되는 것 같아요... 제가 잘못 접근하고 있는 걸까요?

좋은 접근이예요! 그 방식을 LottoNumberGenerator에 적용하면 좋을 것 같았어요ㅎ "랜덤하게" 숫자를 뽑는 친구와 "커스텀 확률로" 숫자를 뽑는 친구와 마찬가지로 "정해진" 숫자를 뽑는 친구를 만들면 테스트할 때 충분히 활용할 수 있지 않을까요? `Math.random()`을 갈아끼우는 거죠. 멍구는 "랜덤"을 테스트하려고 했나요, 랜덤을 활용한 "로직"을 검증하려고 했나요?

> 도메인 캡슐화 때문에 private으로 정의해야한다고 막연히 생각했습니다. 도메인 내부 규칙이 외부에 노출되게 되는데 상수는 불변이기 때문에 노출되어도 상관 없을까요?

대체로 객체의 상태를 private으로 관리하기 때문에 객체 안의 상수 또한 private으로 두어야 한다고 보곤 하는데요. 객체의 상태를 private로 관리해야 하는 이유가 무엇일까요? 외부에 노출되는 것에 어떠한 문제가 있는 걸까요?

> 또, 로또 번호 개수(6개)는 로또 번호를 개수대로 생성하는 번호생성기에서 사용하고 있기도 한데요, 이처럼 여러곳에서 사용되는 상수라면 도메인 규칙에 관련되어있지만 공통으로 관리하는 것이 좋을까요? 현재 상수 클래스에서 관리하고 있는 모든 상수가 Lotto 말고도 다른 클래스에서 사용하고 있습니다(검증 등)

상수는 대체로 그 상수와 가장 연관된 객체가 있어요. 저는 주로 객체에 응집하여 관리하곤 합니다. 그리고 연관된 객체가 없는 애매한 상황은 객체가 필요한 순간이기도 합니다.

yuyoungrhee committed:
refactor(Lotto): 가독성 향상, 필요없는 기능 삭제

- 수직거리를 고려하여 메서드 순서를 배치
- Lotto에는 정렬 기능이 필요 없기 때문에 sortNumbers를 삭제, OutputView에서 정렬하도록 변경
- containsNumber 메서드명을 contains로 변경

yuyoungrhee committed:
refactor(LottoNumber): 로또 번호 원시값 포장

하나의 로또 번호에 대해서도 객체로 만들고, 로또 번호 범위에 대한 책임을 가짐.

yuyoungrhee committed:
refactor(Money): 필요없는 공백 삭제

yuyoungrhee committed:
refactor(NumbersGenerator): 숫자 생성기 인터페이스 활용

yuyoungrhee committed:
refactor(LottoShop): 로또 가격을 LottoShop에서 관리하도록 책임을 Money에서 가져옴

yuyoungrhee committed:
refactor(MatchRank): 코드 변수명 변경

코드 변경으로 남아있던 statistics 변수명을 rank로 변경,

yuyoungrhee committed:
fix(LottoNumber): toString 재정의

toString을 재정의하여 객체를 출력할 때 주소가 아니라 내부 int를 반환하도록 함

yuyoungrhee committed:
refactor(MatchRank): 출력 로직을 view로 옮김

yuyoungrhee committed:
refactor(Money): Money 객체를 통해 돈자체에 대한 유효성 검증을 하도록 함

yuyoungrhee committed:
refactor(InputView): 컨벤션에 따른 수정

yuyoungrhee committed:
refactor: MatchResult 삭제에 따른 리팩터링 진행

yuyoungrhee commented:
안녕하세요 이프! 개인적인 일정이 있어 이번 코멘트의 모든 것을 반영하지는 못했지만 설계 면으로 최대한 반영해봤습니다. 테스트는 가능한 한 빠르게 추가하겠습니다.

## 코멘트에 대한 답변

### 테스트를 위한 getter?
> 멍구는 getter를 왜 열어두면 안된다고 생각하시나요? 말씀하셨듯 테스트를 작성할 때 편하지 않으셨나요??

도메인의 캡슐화와 관련되어서입니다. 내부 구현을 외부에서는 몰라야하는데 getter를 열어둠으로써 내부 구현을 알리게 되고, 이를 외부에서 getter를 통해 사용할 수 있게 됩니다. 즉 getter를 통해 **외부에서 해당 도메인을 사용하여 비즈니스 로직을 수행하는 것이 문제**라고 할 수 있습니다.

따라서 출력용, 디버깅용으로는 일부 사용이 허용될 수 있을 것 같습니다. 그래서 꼭 필요한 경우에는 테스트에도 사용할 수 있을 것 같고요.. 

그런데 이번 리팩토링을 하면서 결국 Lotto 정렬 테스트를 하지 않게 되었고, 테스트에서 getter를 사용하지 않게 되었어요! 테스트를 통해 내부 구현을 확인해야해서 getter가 필요하다면, 꼭 필요한 상황일 수 도 있지만 사실 설계가 잘못된 경우일 수 있습니다. **getter 없이도 테스트 할 수 있는 방향**으로 더 고민을 해보는 것이 필요할 것 같아요.

### DTO의 사용과 필요성
> 도메인에 각 계층이 영향을 받는다는 부분이 잘 이해되지 않는데 이부분 자세히 설명 주실 수 있나요? ... 단순히 계층이 쪼개어 있기 때문에 DTO를 사용하는 것은 반대하고 싶어요. 지금 미션 규모에서는 DTO가 사실 필요없다고 생각하거든요.

다시 보니 말이 잘 정리가 안 되어 있네요.. 도메인을 주고받으면 도메인의 변경에 View나 Controller가 영향을 받을 수 있다고 생각했습니다. 예를 들어 필드 변경 같은 경우예요. 이프의 코멘트를 바탕으로 요구사항도 다시 읽어보고, MVC 패턴에 대해서 더 고민해보았습니다.

이번 미션에서는 DTO가 필요 없다는 것에 동의합니다. 이전에는 View ↔ Domain 서로를 전혀 몰라야한다. 라는 강박이 있었던 것 같아요. 중요한 것은 “View가 도메인의 내부 구현에 의존하지 않도록 하는 것”이지 View가 도메인을 무조건 참조하지 말라는 말은 아니었습니다.. MVC 패턴을 잘못 이해하고 있었던 것 같아요.

>  **MVC 패턴 기반으로 리팩터링**해 **view 패키지의 객체가 domain 패키지 객체에 의존할 수 있지만, domain 패키지의 객체는 view 패키지 객체에 의존하지 않도록 구현**한다.

또 현재 구현한 LottoDto와 WalletDto는 **도메인 모델과 거의 동일한 구조라면 굳이 DTO를 만들 필요가 없었습니다.**

계층이 쪼개어 있기 때문에 DTO를 사용하는 것을 반대한다는 뜻은 이해가 잘 안되는데, 어떤 의미로 말씀하신 건 지 설명 부탁드려도 될까요?

### 상수 관리에 대하여 (공통 vs 연관된 객체 내부)
> 상수는 대체로 그 상수와 가장 연관된 객체가 있어요. 저는 주로 객체에 응집하여 관리하곤 합니다. 그리고 연관된 객체가 없는 애매한 상황은 객체가 필요한 순간이기도 합니다.

저 같은 경우는 로또 번호 범위와 관련하여 연관된 객체가 없는 상황이었으므로, LottoNumber라는 객체가 필요한 순간이었군요, 전체적으로 이와 관련하여 피드백해주셔서 LottoNumber라는 객체를 추가했습니다!
### 수정 사항
#### 1. 로또 번호를 포장하는 LottoNumber 객체 사용
- 로또 범위 검증 책임, 로또 범위 관련 상수를 관리합니다

#### 2. 로또 개수 계산에 대한 책임 Money -> LottoShop

#### 3. MatchRank에 남아있던 출력 로직 ->View로 이동

#### 4. MatchResult Dto 삭제
- Lotto에서 일치개수, 보너스일치여부를 계산하는 것뿐만 아니라 이 정보를 이용해 MatchRank를 얻어오는 역할까지 수행합니다.

### 5. Lotto 정렬 기능 삭제
OutputView에서 정렬하여 출력하도록 변경, 정렬은 Lotto를 비롯한 비즈니스 로직에서 필요가 없으므로 Lotto에서 책임을 옮겼습니다

### 남은 사항
테스트 추가, LottoStatisticsTest에 있는 테스트 인자에 대한 고민, inputView의 역할에 대한 고민
- Wallet에 대한 테스트를 구상하기가 힘든데, 어떤 것을 확인하면 좋을까요? 어떤 케이스로 확인해야 할까요?

yuyoungrhee committed:
refactor(inputView): inputView 사용 통일을 위해 getMoney 삭제

yuyoungrhee committed:
refactor: 사용하지 않는 객체와 테스트 삭제

yuyoungrhee committed:
refactor(test): 오타 수정 및 테스트 이름 변경

yuyoungrhee committed:
refactor(Profit): 예외 메시지를 상수로 관리

sinb57 commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -0,0 +1,59 @@
+        if (!(o instanceof LottoNumber)) {
+            return false;
+        }
+        LottoNumber lottoNumber = (LottoNumber) o;
```
IntelliJ가 종종 좋은 제안을 해줄 때도 있답니다~
![image](https://github.com/user-attachments/assets/87efd4b4-2753-46b9-8fac-9ea7e5d0ef47)

yuyoungrhee replied:
수정했습니다!equals 사용이 익숙치 않아서 자료를 참고했는데, 옛날 코드들을 참고했던 것 같습니다.

수정 후 코드는 자바 16부터 instanceof 연산자에 대한 pattern matching을 지원하여 가능한 코드로, instanceof 검사와 형변환을 한 번에 처리하는 코드입니다. 개인적으로 가독성도 더 좋아진 것 같습니다.

sinb57 commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
Comparable 구현체에 대해서는 단위 테스트가 필요 없을까요?

yuyoungrhee replied:
음 처음에는 믿고 쓰자는 생각이었는데, **직접 만든 로직은 검증해야할 것 같습니다.** Comparable의 메서드를 오버라이드 하는 방식으로 "재정의"를 하는 방식으로 제가 코드를 작성했기 때문에 테스트가 필요하다고 생각해서 추가했습니다.

sinb57 commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -0,0 +1,59 @@
+        }
+    }
+
+    public static List<LottoNumber> from(List<Integer> numbers) {
+        return numbers.stream()
+                .map(LottoNumber::new)
+                .collect(Collectors.toList());
+    }
```
값이 동일한 객체가 여러개 생성될 수 있는데, 자원을 재사용할 수 있도록 캐싱을 적용해보면 어떨까요?

sinb57 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -11,84 +10,66 @@
 
-    private static boolean isNumberInRage(int number) {
-        return number < LOTTO_RANGE_MINIMUM || number > LOTTO_RANGE_MAXIMUM;
+        return matchRank;
```
변수를 선언하지 않고 곧바로 return 해도 되지 않을까요?

sinb57 commented on src/main/java/lotto/controller/Controller.java:
```diff
@@ -22,18 +27,20 @@ public Controller(InputController inputController, OutputView outputView) {
+        LottoShop lottoShop = new LottoShop(
+                new RandomNumbersGenerator(LOTTO_RANGE_MINIMUM, LOTTO_RANGE_MAXIMUM, LOTTO_SIZE));
+        int purchasedLottoCount = lottoShop.calculateLottoCount(money);
+        outputView.printLottoCount(purchasedLottoCount);
 
-        LottoShop lottoShop = new LottoShop(new LottoNumberGenerator());
         List<Lotto> lottos = lottoShop.buyLottos(money);
```
`LottoShop#buyLottos` 안에서 `LottoShop#calculateLottoCount` 메서드가 사용되는데, 출력을 위해 한번더 호출되네요! 두번 호출할 필요가 있을까요? `LottoShop#buyLottos`의 반환 값의 size를 출력하면 되지 않을까요?

yuyoungrhee replied:
아 너무 좋은 방법인 것 같습니다. 사실 로또개수를 계산하는 매서드를 pricate으로 만들었다가 사용자에게 구매 개수를 보여주기 위해 다시 public으로 열었거든요.. 반영하겠습니다!

sinb57 commented on src/test/java/lotto/domain/LottoShopTest.java:
```diff
@@ -1,28 +1,43 @@
         // given
-        LottoShop lottoShop = new LottoShop(new LottoNumberGenerator());
-        Money money = new Money(LOTTO_PRIZE * 5);
+        Money money = new Money(LOTTO_PRIZE * 5, LOTTO_PRIZE);
+
+        LottoShop lottoShop = new LottoShop(
+                new RandomNumbersGenerator(LOTTO_RANGE_MINIMUM, LOTTO_RANGE_MAXIMUM, LOTTO_SIZE));
```
상수를 가져와서 테스트하셨군요! 왜 상수를 가져다 사용하셨나요?
LOTTO_PRIZE 상수 값을 잘못하여 100원으로 수정해버려도 테스트가 통과할 텐데 괜찮을까요?

yuyoungrhee replied:
비즈니스로직에 관련된 상수값이 바뀌면 테스트도 그에맞게 자동화될 수 있게 해야한다고 생각했습니다. 
 그런데 남겨주신 코멘트에 해당하는 단점도 잇고, 또 해당 테스트같은 경우 해당 금액에 대해서 로또를 몇 장 구매할 수 있을지 확인하는 용도인데, 원래 의도하는 동작이 명확해지지 않는 것 같습니다.

기대하는 명확한 값을 직접 사용하는 것이 이해도 쉽고 테스트하고자 하는 바를 확실하게 전할 수 있다고 생각하게 되어 수정했습니다.

sinb57 commented on src/main/java/lotto/view/InputView.java:
```diff
@@ -1,6 +1,7 @@
 import java.util.Scanner;
+import lotto.common.utill.InputParser;
 
 public class InputView implements AutoCloseable {
```
지난 번에는 생각이 짧아서 AutoCloseable을 보고는 👍을 드렸는데요.
다시 보니, AutoCloseable의 구현체로 만든 이유가 궁금해지네요?!

yuyoungrhee replied:
페어의 의견을 통해 자동으로 Scanner가 닫힐 수 있도록 구현해봤는데요, 관련해서 다시 공부해보니 AutoCloseable은 try-with-resources 구문 내에서만 자동으로 close()가 호출 된다고 합니다. 즉 제가 작성한 구현체는 실질적으로 작동하지 않습니다. 지금 상태에서는 필요하지 않기 때문에 삭제했습니다! 앞으로 코드 하나를 작성하더라도 잘 알아보고 써야겠습니다..🤔

sinb57 reviewed (changes requested):
안녕하세요 멍구~ 피드백 반영해주셔서 감사해요ㅎ 추가 코멘트 및 답변 남겼으니 확인 부탁드려요

---

> 계층이 쪼개어 있기 때문에 DTO를 사용하는 것을 반대한다는 뜻은 이해가 잘 안되는데, 어떤 의미로 말씀하신 건 지 설명 부탁드려도 될까요?

충분한 고찰 없이 겉모습만 DTO를 흉내내는 것을 반대하고자 말씀드렸어요. 비즈니스 로직과 UI 로직을 왜 나누어야 하는지, 그리고 domain과 view를 나누는 것은 어떠한 의미를 지니고 있는지에 대해 생각을 정리하고 본인의 기준을 정립하길 바랬습니다.

yuyoungrhee committed:
refactor: InputView와 InputController의 책임 분리

InputController가 "어떤 데이터를 받을 건지"에만 관여하도록 변경
- 입력 문구 출력을 InputView 책임으로 변경
- 어떤 입력을 받을지 자료형까지 InputView에서 관리하도록 변경

yuyoungrhee committed:
refactor(LottoNumber): equals에서 instancof 패턴매칭을 이용하도록 변경

yuyoungrhee committed:
refactor(LottoNumber): 해시를 사용하도록 캐싱하도록 함

yuyoungrhee committed:
refactor(Lotto): 변수를 선언하지 않고 바로 return하도록 함

yuyoungrhee committed:
refactor(test): 프로덕션코드의 퍼블릭 상수를 사용하지 않고 테스트하도록 변경

yuyoungrhee committed:
test(LottoNumber): 로또 번호 compareTo에 대한 테스트 추가

yuyoungrhee committed:
refactor(InputView): AutoCloseable를 구현하지 않도록 변경

yuyoungrhee committed:
refactor(LottoShop): 로또구매개수를 계산하는 로직을 private으로 변경

컨트롤러에서는 구매한 로또의 size를 계산하여 사용자에게 전달하도록 함

yuyoungrhee commented:
안녕하세요 이프! 늦어서 정말 죄송합니다. 마지막으로 코멘트 최대한 반영했습니다, 개인적으로 도메인 책임에 대한 고민을 충분히 하지 못한 거 같아서 아쉽네요.. wallet에 대한 책임, 그리고 좋은 테스트 방법에 대한 고민이 남은 것 같아요. 시간 날 때 다시 들여다보고 남은 코멘트에 대해서도 스스로 답을 내리고 싶습니다. 좋은 리뷰 남겨주셔서 정말 많이 고민해보는 시간이 됐던 것 같습니다 감사합니다!

sinb57 commented on src/test/java/lotto/domain/LottoNumberTest.java:
```diff
@@ -0,0 +1,43 @@
+        LottoNumber number2 = of(1);
+        LottoNumber number3 = of(2);
+
+        assertThat(number1.compareTo(number3)).isNegative();
```
```suggestion
        assertThat(number1).isLessThan(number2);
```
`compareTo` 메서드를 검증하는 목적을 달성하고 있지만, 다르게 접근해도 좋아보여요.
`compareTo`를 재정의한 이유가 Comparable의 구현체로 만들기 위해 필요한 요소였잖아요.
때문에 Comparable의 구현체로서 원하는 대로 동작하는지 검증하는 것이 좋지 않을까요?
실제로 sort() 과정에서 비교 행위가 사용되기도 하고, 이를 위해서는 객체 간의 대소 비교가 가능해야 하니까
이를 테스트하기 위해 `assertThat(number1).isLessThan(number2);` 검증을 하면 좋겠어요.

sinb57 commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -0,0 +1,67 @@
+
+    private final int number;
+
+    private LottoNumber(int number) {
+        this.number = number;
+    }
+
+    public static LottoNumber of(int number) {
+        validate(number);
+
+        return CACHE.computeIfAbsent(number, LottoNumber::new);
+    }
```
```suggestion
    private LottoNumber(int number) {
        this.number = number;
        validate(number);
    }

    public static LottoNumber of(int number) {
        return CACHE.computeIfAbsent(number, LottoNumber::new);
    }
```
팩토리 메서드가 여럿 있을 때 메서드마다 다른 검증로직을 적용할 것이 아니라면, 객체의 생성자에 배치하는 것을 권장드려요

sinb57 reviewed (approved):
안녕하세요 멍구~ 2단계 미션 진행하느라 고생 많으셨어요!
비즈니스 로직과 UI 로직을 잘 분리해주셨고 테스트도 잘 작성해주셨어요.
로또 미션에서 충분히 학습한 듯 하여 리뷰를 마치고자 Approve & Merge 하도록 하겠습니다!
첫번째 미션을 함께하여 영광이었습니다! 앞으로의 우테코 생활을 응원합니닷!
